runelite-client/src/main/java/net/runelite/client/plugins/grandexchange/GrandExchangePanel.java

```diff
@@ -0,0 +1,625 @@
+package net.runelite.client.plugins.grandexchange;
+
+import java.awt.*;
+import java.util.Arrays;
+import java.util.Objects;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+import java.util.function.Consumer;
+import java.util.function.Function;
+import javax.swing.*;
+import javax.swing.GroupLayout.Alignment;
+import javax.swing.LayoutStyle.ComponentPlacement;
+import javax.swing.border.LineBorder;
+import net.runelite.api.Client;
+import net.runelite.api.GrandExchangeOffer;
+import net.runelite.client.RuneLite;
+import net.runelite.client.game.ItemManager;
+import net.runelite.client.ui.PluginPanel;
+
+/**
+ * @author Robbie, created on 29/10/2017 09:58 AM
+ */
+public class GrandExchangePanel extends PluginPanel
+{
+
+  private Client client;
+
+  private JComponent component;
+  private static GrandExchangePanel instance;
+
+
+  private JLabel[] imageLabels;
+  private ImageIcon[] offerImages;
+  private JLabel[] itemNameLabels;
+  private JLabel[] offerPriceLabels;
+  private JLabel[] statusIndicators;
+  private JProgressBar[] progressBars;
+
+  private GrandExchangePanel()
+  {
+    setMinimumSize(new Dimension(PANEL_WIDTH, PANEL_HEIGHT));
+    setPreferredSize(new Dimension(PANEL_WIDTH, PANEL_HEIGHT));
+    setSize(PANEL_WIDTH, PANEL_HEIGHT);
+    setVisible(true);
+  }
+
+  public static GrandExchangePanel getInstance()
+  {
+    return instance == null
+        ? instance = new GrandExchangePanel()
+        : instance;
+  }
+
+  public void init()
+  {
+    if (component == null) {
+      client = RuneLite.getClient();
+      populate();
+      component = buildPanel();
+      add(component);
+      handleLogic();
+    }
+  }
+
+  private void populate()
+  {
+    this.offerImages = new ImageIcon[8];
+    this.imageLabels = new JLabel[8];
+    this.itemNameLabels = new JLabel[8];
+    this.offerPriceLabels = new JLabel[8];
+    this.statusIndicators = new JLabel[8];
+    this.progressBars = new JProgressBar[8];
+
+    //TODO: via java 8 stream?
+    for (int i = 0; i < offerImages.length; i++) {
+      offerImages[i] = new ImageIcon(ItemManager.getImage(0)); //the icon (dwarf remains)
+      imageLabels[i] = new JLabel(); //holds icon
+      itemNameLabels[i] = new JLabel(); //holds item name
+      offerPriceLabels[i] = new JLabel(); //holds offer price
+      statusIndicators[i] = new JLabel(); //holds quantity bought/sold
+      progressBars[i] = new JProgressBar();
+    }
+
+  }
+
+  private void handleLogic()
+  {
+    ScheduledExecutorService ses = Executors.newSingleThreadScheduledExecutor();
+    ses.scheduleWithFixedDelay(() -> updateUI.accept(client.getGrandExchangeOffers()), 0, 1000, TimeUnit.MILLISECONDS);
```

Comment:
There is an ```@Scheduled``` annotation to schedule periodic tasks. Although it may be better to hook to the offers field being modified in the client than to poll it.